x-pack/plugins/ingest_manager/server/services/setup.ts

```diff
@@ -30,70 +30,93 @@ import { appContextService } from './app_context';
 const FLEET_ENROLL_USERNAME = 'fleet_enroll';
 const FLEET_ENROLL_ROLE = 'fleet_enroll';
 
+// the promise which tracks the setup
+let setupIngestStatus: Promise<void> | undefined;
+// default resolve & reject to guard against "undefined is not a function" errors
+let onSetupResolve = () => {};
+let onSetupReject = (error: Error) => {};
+
 export async function setupIngestManager(
   soClient: SavedObjectsClientContract,
   callCluster: CallESAsCurrentUser
 ) {
-  const [installedPackages, defaultOutput, config] = await Promise.all([
-    // packages installed by default
-    ensureInstalledDefaultPackages(soClient, callCluster),
-    outputService.ensureDefaultOutput(soClient),
-    agentConfigService.ensureDefaultAgentConfig(soClient),
-    ensureDefaultIndices(callCluster),
-    settingsService.getSettings(soClient).catch((e: any) => {
-      if (e.isBoom && e.output.statusCode === 404) {
-        const http = appContextService.getHttpSetup();
-        const serverInfo = http.getServerInfo();
-        const basePath = http.basePath;
-
-        const cloud = appContextService.getCloud();
-        const cloudId = cloud?.isCloudEnabled && cloud.cloudId;
-        const cloudUrl = cloudId && decodeCloudId(cloudId)?.kibanaUrl;
-        const flagsUrl = appContextService.getConfig()?.fleet?.kibana?.host;
-        const defaultUrl = url.format({
-          protocol: serverInfo.protocol,
-          hostname: serverInfo.host,
-          port: serverInfo.port,
-          pathname: basePath.serverBasePath,
-        });
-
-        return settingsService.saveSettings(soClient, {
-          agent_auto_upgrade: true,
-          package_auto_upgrade: true,
-          kibana_url: cloudUrl || flagsUrl || defaultUrl,
-        });
-      }
-
-      return Promise.reject(e);
-    }),
-  ]);
-
-  // ensure default packages are added to the default conifg
-  const configWithDatasource = await agentConfigService.get(soClient, config.id, true);
-  if (!configWithDatasource) {
-    throw new Error('Config not found');
-  }
-  if (
-    configWithDatasource.datasources.length &&
-    typeof configWithDatasource.datasources[0] === 'string'
-  ) {
-    throw new Error('Config not found');
+  // installation in progress
+  if (setupIngestStatus) {
+    await setupIngestStatus;
+  } else {
+    // create the initial promise
+    setupIngestStatus = new Promise((res, rej) => {
+      onSetupResolve = res;
+      onSetupReject = rej;
+    });
   }
-  for (const installedPackage of installedPackages) {
-    const packageShouldBeInstalled = DEFAULT_AGENT_CONFIGS_PACKAGES.some(
-      (packageName) => installedPackage.name === packageName
-    );
-    if (!packageShouldBeInstalled) {
-      continue;
+  try {
```

Comment:
Any reason we can't put this whole try/catch block into the `else` above so we don't need to run all this code again since we now know that it ran/is running? Would speed things up a bit, too.

Reply:
That would only run the setup one time which isn't the current or previous behavior. I'm trying to avoid any changes to the prior flow / guarantees or only make the smallest incremental changes required.

We'll definitely revisit this and other parts of the setup in later PRs but for now I want small conservative changes.

Reply:
Okay, but we were only ever running the entire setup twice because we couldn't track the first.  Now there is no need. Otherwise this change is going to cause the Ingest Manager setup to take longer because we have to await the first call before running the whole of the second.

Reply:
> Okay, but we were only ever running the entire setup twice because we couldn't track the first. Now there is no need.

Then maybe we can remove the second call? Either way, I'm starting with stability then we can move on to speed.

Reply:
We still need the second call, which would be more like a check now, because its purpose is to not let the user interact with Ingest Manager until setup has completed.

Reply:
I think we should talk about this outside this PR. In person or another issue. As far as I can tell, this PR fixes the known bugs and tests are passing. I'd like to take the win and do more in another effort.